File: 0031_Next_Permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(len(nums) - 2, -1, -1):
            if nums[i] < nums[i + 1]:
                p = i
                pv = nums[i]
                break
        else:
            nums = nums.reverse()
            return
        
        for i in range(len(nums) - 1, -1, -1):
            if nums[i] > pv:
                nums[i], nums[p] = nums[p], nums[i]
                l, r = p + 1, len(nums) - 1
                while l < r:
                    nums[l], nums[r] = nums[r], nums[l]
                    l += 1
                    r -= 1
                return

File: 0031_Next_Permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int low_index = nums.size();
        for (int i = nums.size() - 2; i >= 0; --i) {
            if (nums[i] < nums[i+1]) {
                low_index = i;
                break;
            }
        }

        if (low_index == nums.size()) {
            sort(nums.begin(), nums.end());
            return;
        }

        int next_larger_index = low_index + 1;
        for (int i = nums.size() - 1; i > low_index + 1; --i) {
            if (nums[i] < nums[next_larger_index] && nums[i] > nums[low_index]) {
                next_larger_index = i;
            }
        }

        swap(nums[low_index], nums[next_larger_index]);
        sort(nums.begin()+low_index+1, nums.end());
        return;
    }
};
